backup/restore functionality - PVC subset

miguelsorianod opened the pull request:
Merge backup/restore PVC subset new functionality integration branch into master

eguzki commented on pkg/apis/apps/v1alpha1/apimanagerbackup_types.go:
```diff
@@ -0,0 +1,156 @@
+type APIManagerBackupDestination struct {
+	// PersistentVolumeClaim as backup data destination configuration
+	// +optional
+	PersistentVolumeClaim *PersistentVolumeClaimBackupDestination `json:"persistentVolumeClaim,omitempty"`
```
why not actual `PersistentVolumeClaim`

miguelsorianod replied:
Only a subset of functionality is provided

eguzki commented on pkg/apis/apps/v1alpha1/apimanagerrestore_types.go:
```diff
@@ -0,0 +1,104 @@
+type APIManagerRestoreSource struct {
+	// +optional
+	// Restore data soure configuration
+	PersistentVolumeClaim *PersistentVolumeClaimRestoreSource `json:"persistentVolumeClaim,omitempty"`
```
why not actual `PersistentVolumeClaim`

miguelsorianod replied:
Only a subset of functionality is provided

eguzki commented on doc/operator-backup-and-restore.md:
````diff
@@ -0,0 +1,130 @@
+           resources:
+             requests: "10Gi"
+   ```
+1. Wait until APIManagerBackup finishes. You can check this by obtaining
````
`oc wait --for=condition=available --timeout=-1s APIManagerBackup/example-apimanagerbackup-pvc`

miguelsorianod replied:
This is not a DeploymentConfig.
There's no "available" condition / conditions.

eguzki replied:
backup type has condition list. Is not used at all? then shoud be removed?

miguelsorianod replied:
Removed

eguzki commented on pkg/apis/apps/v1alpha1/apimanagerbackup_types.go:
```diff
@@ -0,0 +1,156 @@
+type APIManagerBackupConditionType string
+
+// APIManagerBackupCondition describes the state of an APIManagerBackup at a certain point.
+type APIManagerBackupCondition struct {
```
I have used this same type for other features:

https://github.com/3scale/3scale-operator/blob/capabilities-v2/pkg/common/status_conditions.go#L57

It does not have anything specific for backup or restore

miguelsorianod replied:
:+1:  operator-sdk is also adding that in future versions. If needed that could be used when upgrading

eguzki commented on pkg/backup/apimanager_backup_options.go:
```diff
@@ -0,0 +1,26 @@
+	APIManagerBackupName       string                      `validate:"required"` // Name of the APIManagerBackup CR. NOT the APIManager cr name
+	APIManagerBackupUID        types.UID                   `validate:"required"` // UID of the APIManagerBackup CR
+	APIManagerName             string                      `validate:"required"` // Name of the APIManager CR. NOT the APIManagerBackup cr name
+	APIManager                 *appsv1alpha1.APIManager    `validate:"required"` // Should we make this required?
```
I think it should be required. APIManager must be available.

miguelsorianod replied:
:+1: I'll remove the comment

eguzki commented on pkg/backup/apimanager_backup_options_provider.go:
```diff
@@ -0,0 +1,107 @@
+
+}
+
+func (a *APIManagerBackupOptionsProvider) apiManagerFromName(name string) (*appsv1alpha1.APIManager, error) {
```
called from where??

miguelsorianod replied:
it was used before but not anymore. I'll delete it :+1:

eguzki commented on pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:
```diff
@@ -0,0 +1,534 @@
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerRestoreLogicReconciler) reconcileRestoreFromS3Source() (reconcile.Result, error) {
```
:shrug:

miguelsorianod replied:
:+1:

eguzki commented on pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:
```diff
@@ -0,0 +1,534 @@
+		return reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+	}
+
+	// TODO We do not reconcile ownerReference or labels nor annotations
```
Change comment

```
// Jobs ownerReference or labels nor annotations not reconciled
// Jobs are one-shot so there's not much point on making updates to them
```

miguelsorianod replied:
:+1:

eguzki commented on pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:
```diff
@@ -0,0 +1,534 @@
+	}
+
+	// https://godoc.org/k8s.io/apimachinery/pkg/runtime/serializer#CodecFactory
+	codecFactory := k8sserializer.NewCodecFactory(r.Scheme())
```
Why not simply json Marshall/Unmarshall? 

Add a comment explaining why some weird serializer and deserializer is used

miguelsorianod replied:
This is just Kubernetes way of deserializing objects into a runtime object (provided by apimachinery). The own decoder.Decode method in apimachinery contains information about it using the kind, version (from scheme) to deserialize to a runtime object

eguzki replied:
This is comment should be in the code and not lost in github PR

eguzki commented on pkg/restore/apimanager_restore.go:
```diff
@@ -0,0 +1,418 @@
+var secretsToRestore map[string]string = map[string]string{
+	"SystemSMTP": "system-smtp",
+	"SystemSeed": "system-seed",
+	// "SystemDatabase":      "system-database", // We decided this is done by the user. Reason for that is that he might want to restore it in another place
```
These type of comments are useful for the code?

miguelsorianod replied:
I think so. It explains why this subset of secrets/configmaps have not been applied

miguelsorianod commented on deploy/role.yaml:
```diff
@@ -66,6 +62,12 @@ rules:
   - pods
```
:+1:

miguelsorianod commented on deploy/role.yaml:
```diff
@@ -77,6 +79,8 @@ rules:
   - apps.3scale.net
   resources:
   - '*'
+  - apimanagerbackups
```
This was automatically added by operator-sdk when creating the CRDs

eguzki replied:
and apimanager is not automatically added? It should be consistent, either all of them or none.

miguelsorianod replied:
:+1: added apimanager and removed wildcard

eguzki replied:
I do not see apimanagers added

miguelsorianod replied:
See my last comment. I did but tests started to fail due to it seems other (sub)resources are needed so I reverted it to * and removed the additional ones to leave it consistent

miguelsorianod replied:
Opened Jira issue THREESCALE-2998 to investigate why errors are received when replacing `*` by specific custom resource names

miguelsorianod commented:
Applied changes and answered comments

eguzki commented:
comments left

miguelsorianod commented:
Changes applied

eguzki commented:
LGTM

Fix failing tests and let's merge

miguelsorianod commented:
Fixed failing tests: added back the wildcard permission on apps.3scale.net, the removal of it was causing errors creating some objects https://app.circleci.com/pipelines/github/3scale/3scale-operator/2061/workflows/749dbf90-69c4-40d6-821b-2c8d87a2fc2f/jobs/13598
